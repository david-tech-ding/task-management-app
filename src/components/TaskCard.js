import React, { useState } from "react";
import "../styles/taskcard.css";
import Status from "./Status";

const TaskCard = ({ title, priorityLevel, details, dueDate, status }) => {
  const [comment, setComment] = useState("");

  const handleChange = (e) => {
    e.preventDefault();
    setComment(e.target.value);
  };
  const handleSaveComment = (e) => {
    e.preventDefault();
  };

  return (
    <div className="task-card-page">
      <div className="task-card-grid">
        <div className={`task-card ${priorityLevel}`}>
          <div className="task-card_title">{title}</div>
          <div className="task-card_priority_level">
            {priorityLevel}&nbsp;Priority
          </div>
          <div className="task-card_due_date">To be completed by {dueDate}</div>
          <div className="task-card_details">{details}</div>
          <Status className="task-card_status" status={status} />
          <form className="task-card_comments" onSubmit={handleSaveComment}>
            <input
              type="text"
              placeholder="Write a comment..."
              value={comment}
              onChange={handleChange}
            />
            <button type="submit">Save</button>
          </form>
        </div>
      </div>
    </div>
  );
};

export default TaskCard;
