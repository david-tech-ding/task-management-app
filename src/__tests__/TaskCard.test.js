import React from "react";
import { render, screen } from "@testing-library/react";
import axios from "axios";
import TaskCard from "../components/TaskCard";

describe("renders", () => {
  const validProps = {
    title: "test",
    priorityLevel: "test",
    details: "test",
    dueDate: "03 August 2023",
    status: "test",
    id: 1,
    user: "test",
    assignedBy: "test",
    assignTo: 1,
    usersList: [{ firstName: "test", lastName: "test", id: 1 }],
  };

  it("renders correctly", () => {
    const { asFragment } = render(
      <TaskCard
        title={validProps.title}
        priorityLevel={validProps.priorityLevel}
        details={validProps.details}
        dueDate={validProps.dueDate}
        status={validProps.status}
        id={validProps.id}
        user={validProps.user}
        assignedBy={validProps.assignedBy}
        assignTo={validProps.assignTo}
        usersList={validProps.usersList}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("conditionally renders a delete button if userName === assignedBy", () => {
    const testuser = {
      userName: "test",
    };
    render(
      <TaskCard
        title={validProps.title}
        priorityLevel={validProps.priorityLevel}
        details={validProps.details}
        dueDate={validProps.dueDate}
        status={validProps.status}
        id={validProps.id}
        user={testuser}
        assignedBy={validProps.assignedBy}
        assignTo={validProps.assignTo}
        usersList={validProps.usersList}
      />
    );

    const deleteButton = screen.getByText(/delete/i);

    expect(deleteButton).toBeInTheDocument();
  });
});

describe("get comments", () => {
  jest.mock("axios");
  const validProps = {
    title: "test",
    priorityLevel: "test",
    details: "test",
    dueDate: "03 August 2023",
    status: "test",
    id: 1,
    user: "test",
    assignedBy: "test",
    assignTo: 1,
    usersList: [{ firstName: "test", lastName: "test", id: 1 }],
  };

  it("Should return comments list", () => {
    const mockData = ["test comment", "test comment 2"];
    axios.get.mockResolvedValueOnce(mockData);
    render(
      <TaskCard
        title={validProps.title}
        priorityLevel={validProps.priorityLevel}
        details={validProps.details}
        dueDate={validProps.dueDate}
        status={validProps.status}
        id={validProps.id}
        user={validProps.user}
        assignedBy={validProps.assignedBy}
        assignTo={validProps.assignTo}
        usersList={validProps.usersList}
      />
    );

    expect(axios.get).toHaveBeenCalledWith(`/comments/${validProps.id}`);
  });
});
