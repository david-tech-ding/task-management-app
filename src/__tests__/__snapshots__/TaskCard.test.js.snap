// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders renders correctly 1`] = `
<DocumentFragment>
  <div
    class="task-card-page"
  >
    <div
      class="task-card-grid"
    >
      <div
        class="task-card test"
      >
        <div
          class="task-card_title"
        >
          test
        </div>
        <div
          class="task-card_priority_level"
        >
          test Priority
        </div>
        <div
          class="task-card_due_date"
        >
          Due on 03 August 2023
        </div>
        <div
          class="task-card_details"
        >
          test
        </div>
        <div
          class="task-card_status"
        >
          Status: test
        </div>
        <form
          class="task-card_form"
        >
          <select
            class="status-select"
            id="status-select"
          >
            <option
              value="Not Started"
            >
              Not Started
            </option>
            <option
              value="In Progress"
            >
              In Progress
            </option>
            <option
              value="Complete"
            >
              Complete
            </option>
          </select>
          <button
            class="status-button"
            type="button"
          >
            Change Status
          </button>
          <div>
            <h3>
              Comments
            </h3>
            <ul />
          </div>
          <input
            class="comments-input"
            placeholder="Write a comment..."
            type="text"
            value=""
          />
          <button
            class="comments-button"
            type="submit"
          >
            Save
          </button>
        </form>
        <div>
          <h3>
            Assign To
          </h3>
          <select
            class="assigned-user_select"
            id="assigned-user_select"
          >
            <option
              value="1"
            >
              test test ()
            </option>
          </select>
          <button
            class="save-assigned-user-button"
            type="button"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
